File: frontend/lib/api/planner.ts
// Planner API client functions

const API_BASE = "";

export interface PlanRequest {
    message: string;
    context?: {
        timezone?: string;
        today?: string;
    };
}

export interface PlanResponse {
    session_id: string;
    plan: {
        project: {
            title: string;
            description?: string;
        };
        tasks: Array<{
            title: string;
            description?: string;
            due_at: string;
            evidence_type?: string;
            tags?: string[];
        }>;
        rationale?: string;
    };
}

export interface CommitRequest {
    session_id: string;
    plan: PlanResponse['plan'];
}

export interface CommitResponse {
    project_id: string;
    task_ids: string[];
}

export async function planTask(
    message: string,
    token: string
): Promise<PlanResponse> {
    const response = await fetch(`${API_BASE}/api/planner/plan`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({
            message,
            context: {
                timezone: "Asia/Shanghai",
                today: new Date().toISOString().split('T')[0],
            },
        }),
    });

    if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || "规划生成失败");
    }

    return response.json();
}

export async function commitPlan(
    sessionId: string,
    planData: PlanResponse['plan'],
    token: string
): Promise<CommitResponse> {
    const response = await fetch(`${API_BASE}/api/planner/commit`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({
            session_id: sessionId,
            plan: planData,
        }),
    });

    if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || "计划提交失败");
    }

    return response.json();
}

File: app/schemas/planner.py
"""Planner Pydantic schemas."""
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


class PlanContext(BaseModel):
    """Context information for planning."""
    timezone: str = "Asia/Shanghai"
    today: str = Field(..., description="Current date in YYYY-MM-DD format")
    active_projects: Optional[List[dict]] = None
    constraints: Optional[dict] = None


class PlanRequest(BaseModel):
    """Request to generate a plan."""
    message: str = Field(..., min_length=1, max_length=2000, description="User's natural language input")
    context: Optional[PlanContext] = None


class PlannedTask(BaseModel):
    """A single planned task."""
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    due_at: str = Field(..., description="ISO 8601 datetime with timezone")
    evidence_type: str = Field(default="none", pattern="^(none|text|number|image)$")
    tags: Optional[List[str]] = None


class PlannedProject(BaseModel):
    """Planned project information."""
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None


class PlanResponse(BaseModel):
    """Response containing the generated plan."""
    session_id: str
    plan: dict = Field(..., description="Structured plan with project and tasks")
    
    class Config:
        json_schema_extra = {
            "example": {
                "session_id": "uuid-here",
                "plan": {
                    "project": {
                        "title": "微积分学习计划",
                        "description": "3个月完成微积分基础"
                    },
                    "tasks": [
                        {
                            "title": "完成极限章节",
                            "description": "阅读教材 + 做练习题50道",
                            "due_at": "2026-02-15T23:59:59+08:00",
                            "evidence_type": "number"
                        }
                    ],
                    "rationale": "分为6个两周周期"
                }
            }
        }


class CommitRequest(BaseModel):
    """Request to commit a plan and create project + tasks."""
    session_id: str
    plan: dict = Field(..., description="The plan data to commit")


class CommitResponse(BaseModel):
    """Response after successfully committing a plan."""
    project_id: str
    task_ids: List[str]
    
    class Config:
        json_schema_extra = {
            "example": {
                "project_id": "proj-uuid",
                "task_ids": ["task1-uuid", "task2-uuid", "task3-uuid"]
            }
        }
